WordPress/src/main/java/org/wordpress/android/ui/JetpackConnectionWebViewClient.java

```diff
@@ -0,0 +1,137 @@
+import org.wordpress.android.fluxc.store.AccountStore;
+import org.wordpress.android.fluxc.store.SiteStore;
+import org.wordpress.android.login.LoginMode;
+import org.wordpress.android.ui.accounts.LoginActivity;
+import org.wordpress.android.ui.prefs.AppPrefs;
+import org.wordpress.android.util.AppLog;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+
+class JetpackConnectionWebViewClient extends WebViewClient {
+
+    private static final String LOGIN_PATH = "/wp-login.php";
+    private static final String ADMIN_PATH = "/wp-admin/admin.php";
+    private static final String REDIRECT_PARAMETER = "redirect_to=";
+    private static final int REQUEST_CODE = 1;
+    private static final String WORDPRESS_COM_HOST = "wordpress.com";
+    private static final String LOG_IN_PATH = "/log-in";
+    private static final String JETPACK_PATH = "/jetpack";
+    private static final String WORDPRESS_COM_PREFIX = "https://wordpress.com";
+    private static final Uri JETPACK_DEEPLINK_URI = Uri.parse(WPWebViewActivity.JETPACK_CONNECTION_DEEPLINK);
+    private static final String REDIRECT_PAGE_STATE_ITEM = "redirectPage";
+    private static final String FLOW_FINISHED = "FLOW_FINISHED";
+
+    private Activity activity;
+    private final AccountStore accountStore;
+    private final SiteStore mSiteStore;
+
+    private String redirectPage;
+    private boolean flowFinished = false;
+
+    JetpackConnectionWebViewClient(Activity activity, AccountStore accountStore, SiteStore mSiteStore) {
+        this.activity = activity;
+        this.accountStore = accountStore;
+        this.mSiteStore = mSiteStore;
+    }
+
+    private void loginToWPCom(WebView view, SiteModel site) {
+        String authenticationURL = WPWebViewActivity.getSiteLoginUrl(site);
+        String postData = WPWebViewActivity.getAuthenticationPostData(authenticationURL, redirectPage, site.getUsername(), site.getPassword(), accountStore.getAccessToken());
+        view.postUrl(authenticationURL, postData.getBytes());
+    }
+
+    private String extractRedirect(String stringUrl) throws UnsupportedEncodingException {
+        int from = stringUrl.indexOf(REDIRECT_PARAMETER) + REDIRECT_PARAMETER.length();
+        int to = stringUrl.indexOf("&", from);
+        String redirectUrl;
+        if (to > from) {
+            redirectUrl = stringUrl.substring(from, to);
+        } else {
+            redirectUrl = stringUrl.substring(from);
+        }
+        if (redirectUrl.startsWith(JETPACK_PATH)) {
+            redirectUrl = WORDPRESS_COM_PREFIX + redirectUrl;
+        }
+        return URLDecoder.decode(redirectUrl, WPWebViewActivity.ENCODING_UTF8);
+    }
+
+    @Override
+    public boolean shouldOverrideUrlLoading(WebView view, String stringUrl) {
+        try {
+            final Uri url = Uri.parse(stringUrl);
+            final SiteModel site = mSiteStore.getSiteByLocalId(AppPrefs.getSelectedSite());
+            final String loadedHost = url.getHost();
```

Comment:
I think we need to add some null-checking on `loadedHost` and `loadedPath` because there are cases where `getPath()` and `getHost()`  can return null. It's a rare possibility here, but since the user can click on numerous external links during this flow it's worth checking for.